sopel/modules/safety.py

```diff
@@ -212,80 +250,167 @@ def url_handler(bot: Sopel, trigger: Trigger):
-            r = requests.post(
-                VT_API_URL,
-                data={"url": url},
+    attempts = 5
+    requested = False
+    while attempts > 0:
+        attempts -= 1
+        try:
+            r = requests.get(
+                VT_API_URL + "/" + url_id,
                 headers={"x-apikey": bot.settings.safety.vt_api_key},
             )
+            if r.status_code == 200:
+                vt_data = r.json()
+                last_analysis = vt_data["data"]["attributes"]["last_analysis_stats"]
+                # VT returns 200s for recent submissions before scan results are in...
+                if not requested or sum(last_analysis.values()) > 0:
+                    break
+            elif not requested and r.status_code == 404:
+                # Not analyzed - submit new
+                LOGGER.debug("[VirusTotal] No scan for %r, requesting", safe_url)
+                r = requests.post(
+                    VT_API_URL,
+                    data={"url": url},
+                    headers={"x-apikey": bot.settings.safety.vt_api_key},
+                )
+                requested = True
+                sleep(2)  # Scans seem to take ~5s minimum, so add 2s
+        except requests.exceptions.RequestException:
+            # Ignoring exceptions with VT so domain list will always work
+            LOGGER.debug(
+                "[VirusTotal] Error obtaining response for %r", safe_url, exc_info=True
+            )
             return None
-        r.raise_for_status()
-        vt_data = r.json()
-    except requests.exceptions.RequestException:
-        # Ignoring exceptions with VT so domain list will always work
-        LOGGER.debug(
-            "[VirusTotal] Error obtaining response for %r", safe_url, exc_info=True
-        )
-    except json.JSONDecodeError:
-        # Ignoring exceptions with VT so domain list will always work
-        LOGGER.debug(
-            "[VirusTotal] Malformed response (invalid JSON) for %r",
-            safe_url,
-            exc_info=True,
-        )
-    fetched = time.time()
-    last_analysis = vt_data["data"]["attributes"]["last_analysis_stats"]
+        except json.JSONDecodeError:
+            # Ignoring exceptions with VT so domain list will always work
+            LOGGER.debug(
+                "[VirusTotal] Malformed response (invalid JSON) for %r",
+                safe_url,
+                exc_info=True,
+            )
+            return None
+        sleep(3)
+    else:  # Still no results
+        LOGGER.debug("[VirusTotal] Scan failed or unfinished for %r", safe_url)
+        return None
+    fetched = datetime.now(timezone.utc)
     # Only count strong opinions (ignore suspicious/timeout/undetected)
     result = {
         "positives": last_analysis["malicious"],
         "total": last_analysis["malicious"] + last_analysis["harmless"],
         "fetched": fetched,
-        "virustotal_data": vt_data,
+        # Subject to change formats!
```

Comment:
Don't quite understand what this comment is telling me.

Reply:
If VT changes how it returns data, the structure of that dict will change, with or without a sopel version change.

Reply:
We should have a comment like this next to every use of API data in any plugin, then.

Of course the API data format _could_ change, but it _won't_ if VT is actually serious about the API version number.